libraries/chain_kv/include/b1/session/session.hpp

```diff
@@ -0,0 +1,1061 @@
+        }
+    }
+    
+    // For all the keys that were found in one of the session caches, read them from the persistent data store
+    // and write them into our cache.
+    if (m_impl->backing_data_store)
+    {
+        auto bds_kvs = m_impl->backing_data_store->read(not_found);
+        if (!bds_kvs.empty())
+        {
+            m_impl->cache.write(bds_kvs.first);
+        }
+        not_found = std::move(bds_kvs.second);
+        kvs.insert(std::end(kvs), std::begin(bds_kvs.first), std::end(bds_kvs.first));
+    }
+    
+    return std::make_pair(std::move(kvs), std::move(not_found));
+}
+
+// Writes a batch of key_values to the session.
+//
+// \tparam iterable Any type that can be used within a range based for loop and returns key_value instances in its iterator.
+// \param key_values An iterable instance that returns key_value instances in its iterator.
+template <typename persistent_data_store, typename cache_data_store>
+template <typename iterable>
+auto session<persistent_data_store, cache_data_store>::write(const iterable& key_values) -> void
+{
+    if (!m_impl)
+    {
+        return;
+    }
+    
+    // Currently the batch write will just iteratively call the non batch write
+    for (const auto& kv : key_values)
+    {
+        write(kv);
+    }
+}
+
+// Erases a batch of key_values from the session.
+//
+// \tparam iterable Any type that can be used within a range based for loop and returns bytes instances in its iterator.
+// \param keys An iterable instance that returns bytes instances in its iterator.
+template <typename persistent_data_store, typename cache_data_store>
+template <typename iterable>
+auto session<persistent_data_store, cache_data_store>::erase(const iterable& keys) -> void
+{
+    if (!m_impl)
+    {
+        return;
+    }
+    
+    // Currently the batch erase will just iteratively call the non batch erase
+    for (const auto& key : keys)
+    {
+        erase(key);
+    }
+}
+
+template <typename persistent_data_store, typename cache_data_store>
+template <typename data_store, typename iterable>
+auto session<persistent_data_store, cache_data_store>::write_to(data_store& ds, const iterable& keys) const -> void
+{
+    if (!m_impl)
+    {
+        return;
+    }
+    
+    auto results = std::vector<key_value>{};
+    for (const auto& key : keys)
+    {
+        auto* parent = m_impl.get();
+        while (parent)
+        {
+            if (parent->deleted_keys.find(key) != std::end(parent->deleted_keys))
+            {
+                // key/value has been deleted in this session.  Bail out.
+                break;
+            }
+            
+            auto& kv = parent->cache.read(key);
+            if (kv == key_value::invalid)
+            {
+                continue;
+            }
+            
+            results.emplace_back(make_kv(kv.key().data(), kv.key().length(), kv.value().data(), kv.value().length(), ds.allocator()));
+        }
+    }
+    ds.write(results);
```

Comment:
Currently this only pulls in keys that are in the caches.  If I can't find the key in the cache, should I hit the database to pull it in from there?